src/main/views/loadingScreen.ts

```diff
@@ -0,0 +1,116 @@
+    VISIBLE = 1,
+    FADING = 2,
+    HIDDEN = 3,
+}
+
+export class LoadingScreen {
+    private view?: BrowserView;
+    private state: LoadingScreenState;
+    private mainWindow?: BrowserWindow;
+
+    constructor() {
+        this.state = LoadingScreenState.HIDDEN;
+
+        ipcMain.on(LOADING_SCREEN_ANIMATION_FINISHED, this.handleAnimationFinished);
+    }
+
+    /**
+     * Loading Screen
+     */
+
+    setBounds = () => {
+        if (!(this.view && this.mainWindow)) {
+            return;
+        }
+
+        this.view.setBounds(getWindowBoundaries(this.mainWindow));
+    }
+
+    setDarkMode = (darkMode: boolean) => {
+        this.view?.webContents.send(DARK_MODE_CHANGE, darkMode);
+    }
+
+    isHidden = () => {
+        return this.state === LoadingScreenState.HIDDEN;
+    }
+
+    show = () => {
+        if (!this.mainWindow) {
+            return;
+        }
+
+        if (!this.view) {
+            this.create();
+        }
+
+        this.state = LoadingScreenState.VISIBLE;
+
+        if (this.view?.webContents.isLoading()) {
+            this.view.webContents.once('did-finish-load', () => {
+                this.view!.webContents.send(TOGGLE_LOADING_SCREEN_VISIBILITY, true);
+            });
+        } else {
+            this.view!.webContents.send(TOGGLE_LOADING_SCREEN_VISIBILITY, true);
+        }
+
+        if (this.mainWindow.getBrowserViews().includes(this.view!)) {
+            this.mainWindow.setTopBrowserView(this.view!);
+        } else {
+            this.mainWindow.addBrowserView(this.view!);
+        }
+
+        this.setBounds();
+    }
+
+    fade = () => {
+        if (this.view && this.state === LoadingScreenState.VISIBLE) {
+            this.state = LoadingScreenState.FADING;
+            this.view.webContents.send(TOGGLE_LOADING_SCREEN_VISIBILITY, false);
+        }
+    }
+
+    setMainWindow = (mainWindow: BrowserWindow) => {
+        this.mainWindow = mainWindow;
+    }
+
+    private create = () => {
+        const preload = getLocalPreload('desktopAPI.js');
+        this.view = new BrowserView({webPreferences: {
+            preload,
+
+            // Workaround for this issue: https://github.com/electron/electron/issues/30993
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
+            transparent: true,
+        }});
+        const localURL = getLocalURLString('loadingScreen.html');
+        this.view.webContents.loadURL(localURL);
+    }
+
+    private handleAnimationFinished = () => {
+        log.debug('handleLoadingScreenAnimationFinished');
+
+        if (this.view && this.state !== LoadingScreenState.HIDDEN) {
+            this.state = LoadingScreenState.HIDDEN;
+            this.mainWindow?.removeBrowserView(this.view);
+        }
+
+        if (process.env.NODE_ENV === 'test') {
+            app.emit('e2e-app-loaded');
+        }
+    }
+}
+
+const loadingScreen = new LoadingScreen();
+export default loadingScreen;
```

Comment:
Purely out of curiosity, when does the loading screen object get initialized? I suspect it may be caused by an import and the runtime is smart enough to do it only once but not sure. 

Mostly I wonder if this is a commonly established pattern in the JS world since from my low-level tuned brain perspective it feels a bit obscure as I'd expect it to be initialized somewhere in the entry point for the app.

Reply:
So I'm not sure if it's established in the JS world, but I'm aiming to work in the singleton pattern for most objects now since it ensures we encapsulate the logic most relevant to said modules and means we don't end up with as many behemoths like the `windowManager`.

As for how it's imported, I believe webpack is smart enough to handle that. Once the first call of `import` happens for the file it'll run this code and export the object. I haven't done a deep dive into the NodeJS engine to see how it actually works but that's my current understanding.